### PR TITLE
Table Variable rework

### PR DESCRIPTION
**Pull request by lejaokri**
### Description
Change 1

    Add pg_depend, pg_shdepend to ENR for temp tables

Change 2

    (This depends on Change 1)

    Table Variables inside Transactions (engine code changes)

    This commit refactors how Table Variables work inside a Transaction
    especially when there is a rollback. The Table Variable metadata,
    on-disk files, and tuples should be able to survive rollback
    because Table Variables are non-transactional.

    Also, Table Variables primary key, constraint and toast tables should
    be non-transactional as well.

    On high-level there are three main components in this commit:

      1. Metadata:
         Table Variable metadata are added into ENR in order to bypass
         catalog MVCC.

      2. On-disk files:
         Table Variable relfilenode are not added to pending_deletes
         during table creation so it would not be deleted in case of rollback.

      3. Tuples:
         Introduced a new set of MVCC, Vacuum, Toast visibility functions.

 
### Issues Resolved

[List any issues this PR will resolve]
 
### Check List

- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/access/heap/heapam_visibility.c [hunk 1]
```diff
@@ -1792,3 +1796,40 @@ HeapTupleSatisfiesVisibility(HeapTuple tup, Snapshot snapshot, Buffer buffer)
...
+table_tuple_satisfies_vacuum(Relation relation, HeapTuple stup, TransactionId OldestXmin, Buffer buffer)
+{
+	if (table_tuple_satisfies_vacuum_hook)
+		return table_tuple_satisfies_vacuum_hook(relation, stup, OldestXmin, buffer);
+
+	return HeapTupleSatisfiesVacuum(stup, OldestXmin, buffer);
+}
+
+HTSV_Result
+table_tuple_satisfies_vacuum_horizon(Relation relation, HeapTuple htup, Buffer buffer, TransactionId *dead_after)
```
Comment (macwoody) [inline src/backend/access/heap/heapam_visibility.c]: what is the best way to make sure future direct references to HeapTupleSatisfiesVacuumHorizon would be detected automatically ?

Comment (lejaokri) [reply src/backend/access/heap/heapam_visibility.c]: cant think of an efficient way but add a comment to ensure the `table_tuple_satisfies*` API should be called  and add an assert in these functions to catch if table variables are called for these functions

same thing for these 4 functions

```
221 extern bool HeapTupleSatisfiesVisibility(HeapTuple stup, Snapshot snapshot,
222                                          Buffer buffer);
223 extern TM_Result HeapTupleSatisfiesUpdate(HeapTuple stup, CommandId curcid,
224                                           Buffer buffer);
225 extern HTSV_Result HeapTupleSatisfiesVacuum(HeapTuple stup, TransactionId OldestXmin,
226                                             Buffer buffer);
227 extern HTSV_Result HeapTupleSatisfiesVacuumHorizon(HeapTuple stup, Buffer buffer,
228                                                    TransactionId *dead_after);
```

File: src/backend/catalog/heap.c [hunk 2]
```diff
@@ -1067,6 +1068,47 @@ AddNewRelationType(const char *typeName,
...
+			SysCacheGetAttr(TYPEOID, tup, Anum_pg_type_typacl, &isnull);
+			if (!isnull)
+			{
+				ReleaseSysCache(tup);
+				return 1;
+			}
+			ReleaseSysCache(tup);
+		}
+	}
+	return 0;
```
Comment (surendravishnoi) [inline src/backend/catalog/heap.c]: true/false?

File: src/backend/catalog/index.c [hunk 3]
```diff
@@ -2350,7 +2376,8 @@ index_drop(Oid indexId, bool concurrent, bool concurrent_lock_mode)
 	hasexprs = !heap_attisnull(tuple, Anum_pg_index_indexprs,
 							   RelationGetDescr(indexRelation));
 
-	CatalogTupleDelete(indexRelation, &tuple->t_self);
+	if (!ENRdropTuple(indexRelation, tuple))
+		CatalogTupleDelete(indexRelation, &tuple->t_self);
```
Comment (surendravishnoi) [inline src/backend/catalog/index.c:2380]: is this a valid fallback?

File: src/backend/catalog/namespace.c [hunk 4]
```diff
@@ -4166,6 +4166,11 @@ InitTempTableNamespace(void)
 		/* Advance command counter to make namespace visible */
 		CommandCounterIncrement();
 	}
+	else if (sql_dialect == SQL_DIALECT_TSQL && get_namedRelList() != NIL)
+	{
+		/* Cannot wipe out any ENR relations if any */
```
Comment (surendravishnoi) [inline src/backend/catalog/namespace.c:4171]: what if a temp table table is not part of ENR?

File: src/include/utils/rel.h [hunk 5]
```diff
@@ -694,4 +694,10 @@ RelationGetSmgr(Relation rel)
 extern void RelationIncrementReferenceCount(Relation rel);
 extern void RelationDecrementReferenceCount(Relation rel);
 
+#define RelationIsBBFTableVariable(relation) \
```
Comment (surendravishnoi) [inline src/include/utils/rel.h:697]: may be add dialect check here only as it is done at all the places

File: src/backend/utils/misc/queryenvironment.c [hunk 6]
```diff
@@ -317,6 +327,83 @@ bool ENRgetSystableScan(Relation rel, Oid indexId, int nkeys, ScanKey key, List
...
+					Form_pg_index tup = (Form_pg_index) GETSTRUCT((HeapTuple) lfirst(lc));
+					if (indexId == IndexIndrelidIndexId &&
+						tup->indrelid == (Oid)v1)
+					{
+						*tuplist = list_insert_nth(*tuplist, index++, lfirst(lc));
+						*tuplist_flags |= SYSSCAN_ENR_NEEDFREE;
+						found = true;
+					}
+					else if (indexId == IndexRelidIndexId &&
+							tup->indexrelid == (Oid)v1) {
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: remove if not required

File: src/backend/utils/misc/queryenvironment.c [hunk 7]
```diff
@@ -462,7 +549,69 @@ bool ENRgetSystableScan(Relation rel, Oid indexId, int nkeys, ScanKey key, List
...
+	QueryEnvironment *queryEnv = currentQueryEnv;
+	Oid catalog_oid = entry->classid;
+
+	ListCell         *curlc;
+
+	while (queryEnv)
+	{
+		switch (catalog_oid) {
+			/*
+			* pg_depend entry shows relation depends on a given object.
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: duplicate comment?

File: src/backend/utils/misc/queryenvironment.c [hunk 8]
```diff
@@ -494,37 +643,90 @@ static bool _ENR_tuple_operation(Relation catalog_rel, HeapTuple tup, ENRTupleOp
...
 					list_ptr = &enr->md.cattups[ENR_CATTUP_CLASS];
 					lc = list_head(enr->md.cattups[ENR_CATTUP_CLASS]);
 					ret = true;
 				}
 				break;
+			case DependRelationId:
+				{
+					Form_pg_depend tf1 = (Form_pg_depend) GETSTRUCT((HeapTuple)tup);
+					//if ((enr = find_enr(tf1->classid, tf1->objid, tf1->objsubid))) {
+					if ((enr = find_enr(tf1))) {
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c:647]: have we done good testing around this code to make sure we are taking care of all dependencies and types properly

File: src/backend/catalog/index.c (hunk 3 above)
Comment (lejaokri) [reply src/backend/catalog/index.c:2380]: Yes this is valid. At this point we know the index exists and it hasn't been dropped otherwise we would've failed earlier in the function. If index exists, then it has to be in either ENR or Catalog.

File: src/backend/catalog/namespace.c (hunk 4 above)
Comment (lejaokri) [reply src/backend/catalog/namespace.c:4171]: Good point. @timchang514 please test this for temp tables

File: src/backend/utils/misc/queryenvironment.c (hunk 6 above)
Comment (lejaokri) [reply src/backend/utils/misc/queryenvironment.c]: This is still required for Table Variables

File: src/backend/utils/misc/queryenvironment.c (hunk 8 above)
Comment (lejaokri) [reply src/backend/utils/misc/queryenvironment.c:656]: will add more testcases for this for table variables

File: src/include/utils/rel.h (hunk 5 above)
Comment (lejaokri) [reply src/include/utils/rel.h:697]: I remember the reason I didnt do it that way is `tsql_dialect` is in parser.h and we would have to include parser.h in rel.h. On the other hand, rel.h is included everywhere.

File: src/backend/catalog/namespace.c (hunk 4 above)
Comment (time-change) [reply src/backend/catalog/namespace.c:4171]: I've done some testing, seems that are temp tables are still cleaned up after crashes even when temp table is not part of ENR.

File: src/backend/catalog/heap.c (hunk 2 above)
Comment (time-change) [reply src/backend/catalog/heap.c]: Done

File: src/backend/access/heap/heapam_visibility.c (hunk 1 above)
Comment (peipeiyin) [reply src/backend/access/heap/heapam_visibility.c]: It should work if we add parameter `Relation relation` to functions `HeapTupleSatisfies*` and do bashfish hook check inside these functions. Once there is new reference from community in the future, aurora code build will fail and we can fix the community merge accordingly.

Comment (lejaokri) [reply src/backend/access/heap/heapam_visibility.c]: Thank you. As discussed, I've added a new parameter `Relation relation` inside the `HeapTupleSatisfies*` functions and use the `relation` parameter to detect if it is a Table Variable.

File: src/backend/utils/misc/queryenvironment.c (hunk 8 above)
Comment (time-change) [reply src/backend/utils/misc/queryenvironment.c:647]: tests in https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1440 cover types and dependencies

File: src/backend/access/transam/xact.c [hunk 9]
```diff
@@ -359,7 +359,6 @@ static void ShowTransactionStateRec(const char *str, TransactionState state);
 static const char *BlockStateAsString(TBlockState blockState);
 static const char *TransStateAsString(TransState state);
 
-
```
Comment (surendravishnoi) [inline src/backend/access/transam/xact.c]: not required

File: src/backend/catalog/dependency.c [hunk 10]
```diff
@@ -709,7 +709,7 @@ findDependentObjects(const ObjectAddress *object,
 				 * interesting anymore.  We test this by checking the
 				 * pg_depend entry (see notes below).
 				 */
-				if (!systable_recheck_tuple(scan, tup))
+				if (!scan->enr && !systable_recheck_tuple(scan, tup))
```
Comment (surendravishnoi) [inline src/backend/catalog/dependency.c]: what is the purpose of this check here? shouldn't we return in the beginning of function itself?

File: src/backend/catalog/dependency.c [hunk 11]
```diff
@@ -898,7 +898,7 @@ findDependentObjects(const ObjectAddress *object,
 		 * if the pg_depend tuple we are looking at is still live. (If the
 		 * object got deleted, the tuple would have been deleted too.)
 		 */
-		if (!systable_recheck_tuple(scan, tup))
+		if (!scan->enr && !systable_recheck_tuple(scan, tup))
```
Comment (surendravishnoi) [inline src/backend/catalog/dependency.c]: same as above

File: src/backend/executor/nodeModifyTable.c [hunk 12]
```diff
@@ -1503,7 +1503,7 @@ ExecDelete(ModifyTableContext *context,
 
 	if (resultRelInfo->ri_TrigDesc &&
 		resultRelInfo->ri_TrigDesc->trig_delete_instead_statement &&
-		sql_dialect == SQL_DIALECT_TSQL &&
```
Comment (surendravishnoi) [inline src/backend/executor/nodeModifyTable.c:1506]: unnecessary change

File: src/backend/parser/parse_relation.c [hunk 13]
```diff
@@ -804,7 +804,7 @@ scanRTEForColumn(ParseState *pstate, RangeTblEntry *rte,
 		refname = downcase_identifier(colname, strlen(colname), false, false);
 		reflen = strlen(refname);
 	}
-
```
Comment (surendravishnoi) [inline src/backend/parser/parse_relation.c]: unnecessary change

File: src/backend/utils/misc/queryenvironment.c (hunk 6 above)
Comment (surendravishnoi) [reply src/backend/utils/misc/queryenvironment.c]: I meant remove commented out code

File: src/backend/catalog/dependency.c (hunk 10 above)
Comment (lejaokri) [reply src/backend/catalog/dependency.c]: systable_recheck_tuple cannot handle enr scans because enr scans do not use the tuple slot. I've moved it in the function

File: src/backend/utils/misc/queryenvironment.c (hunk 6 above)
Comment (lejaokri) [reply src/backend/utils/misc/queryenvironment.c]: Done

**Review by surendravishnoi — APPROVED** [review]
